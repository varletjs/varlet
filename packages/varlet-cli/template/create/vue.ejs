<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
