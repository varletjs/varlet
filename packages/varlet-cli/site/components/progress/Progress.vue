<template>
  <div :class="n()">
    <div :class="n('linear')" v-if="mode === 'linear'">
      <div :class="n('linear-block')" :style="{ height: toSizeUnit(lineWidth) }">
        <div :class="n('linear-background')" v-if="track" :style="{ background: trackColor }"></div>
        <div
          :class="classes(n('linear-certain'), [ripple, n('linear-ripple')])"
          :style="{ background: color, width: linearProps.width }"
        ></div>
      </div>
      <div :class="classes(n('linear-label'), [labelClass, labelClass])" v-if="label">
        <slot>
          {{ linearProps.roundValue }}
        </slot>
      </div>
    </div>

    <div :class="n('circle')" v-if="mode === 'circle'" :style="{ width: toSizeUnit(size), height: toSizeUnit(size) }">
      <svg :class="n('circle-svg')" :style="{ transform: `rotate(${rotate - 90}deg)` }" :viewBox="circleProps.viewBox">
        <circle
          v-if="track"
          :class="n('circle-background')"
          :cx="multiplySizeUnit(size, 0.5)"
          :cy="multiplySizeUnit(size, 0.5)"
          :r="circleProps.radius"
          fill="transparent"
          :stroke-width="toSizeUnit(lineWidth)"
          :style="{
            strokeDasharray: circleProps.perimeter,
            stroke: trackColor,
          }"
        ></circle>
        <circle
          :class="n('circle-certain')"
          :cx="multiplySizeUnit(size, 0.5)"
          :cy="multiplySizeUnit(size, 0.5)"
          :r="circleProps.radius"
          fill="transparent"
          :stroke-width="toSizeUnit(lineWidth)"
          :style="{
            strokeDasharray: circleProps.strokeDasharray,
            stroke: color,
          }"
        ></circle>
      </svg>

      <div :class="classes(n('circle-label'), [labelClass, labelClass])" v-if="label">
        <slot>
          {{ circleProps.roundValue }}
        </slot>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue'
import { props } from './props'
import { toNumber } from '@varlet/shared'
import { toSizeUnit, multiplySizeUnit, toPxNum } from '../utils/elements'
import { createNamespace } from '../utils/components'

const { n, classes } = createNamespace('progress')

export default defineComponent({
  name: 'VarProgress',
  props,
  setup(props) {
    const linearProps = computed(() => {
      const value = toNumber(props.value)
      const width = value > 100 ? 100 : value
      const roundValue = value > 100 ? 100 : Math.round(value)

      return {
        width: `${width}%`,
        roundValue: `${roundValue}%`,
      }
    })

    const circleProps = computed(() => {
      const { size, lineWidth, value } = props
      const viewBox = `0 0 ${toPxNum(size)} ${toPxNum(size)}`
      const roundValue = toNumber(value) > 100 ? 100 : Math.round(toNumber(value))
      const radius = (toPxNum(size) - toPxNum(lineWidth)) / 2
      const perimeter = 2 * Math.PI * radius
      const strokeDasharray = `${(roundValue / 100) * perimeter}, ${perimeter}`

      return {
        viewBox,
        radius,
        strokeDasharray,
        perimeter,
        roundValue: `${roundValue}%`,
      }
    })
    return {
      n,
      classes,
      toSizeUnit,
      multiplySizeUnit,
      linearProps,
      circleProps,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import './progress';
</style>
