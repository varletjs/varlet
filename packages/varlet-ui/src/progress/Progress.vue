<template>
  <div :class="n()">
    <div :class="n('linear')" v-if="mode === 'linear'">
      <div :class="n('linear-block')" :style="{ height: `${lineWidth}px` }" v-bind="$attrs">
        <div :class="n('linear-background')" v-if="track" :style="{ background: trackColor }"></div>
        <div
          :class="classes(n('linear-certain'), [ripple, n('linear-ripple')])"
          :style="{ background: color, width: linearProps.width }"
        ></div>
      </div>
      <div :class="n('linear-label')" v-bind="$attrs" v-if="label">
        <slot>
          {{ linearProps.roundValue }}
        </slot>
      </div>
    </div>

    <div :class="n('circle')" v-if="mode === 'circle'" :style="{ width: `${size}px`, height: `${size}px` }">
      <svg :class="n('circle-svg')" :style="{ transform: `rotate(${rotate - 90}deg)` }" :viewBox="circleProps.viewBox">
        <circle
          v-if="track"
          :class="n('circle-background')"
          :cx="size / 2"
          :cy="size / 2"
          :r="circleProps.radius"
          fill="transparent"
          :stroke-width="lineWidth"
          :style="{
            strokeDasharray: circleProps.perimeter,
            stroke: trackColor,
          }"
        ></circle>
        <circle
          :class="n('circle-certain')"
          :cx="size / 2"
          :cy="size / 2"
          :r="circleProps.radius"
          fill="transparent"
          :stroke-width="lineWidth"
          :style="{
            strokeDasharray: circleProps.strokeDasharray,
            stroke: color,
          }"
        ></circle>
      </svg>

      <div :class="n('circle-label')" v-if="label" v-bind="$attrs">
        <slot>
          {{ circleProps.roundValue }}
        </slot>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue'
import { props } from './props'
import { toNumber } from '../utils/shared'
import { createNamespace } from '../utils/components'

const { n, classes } = createNamespace('progress')

export default defineComponent({
  name: 'VarProgress',
  inheritAttrs: false,
  props,
  setup(props) {
    const linearProps = computed(() => {
      const value = toNumber(props.value)
      const width = value > 100 ? 100 : value
      const roundValue = value > 100 ? 100 : Math.round(value)

      return {
        width: `${width}%`,
        roundValue: `${roundValue}%`,
      }
    })

    const circleProps = computed(() => {
      const { size, lineWidth, value } = props
      const viewBox = `0 0 ${size} ${size}`
      const roundValue = toNumber(value) > 100 ? 100 : Math.round(toNumber(value))
      const radius = (size - toNumber(lineWidth)) / 2
      const perimeter = 2 * Math.PI * radius
      const strokeDasharray = `${(roundValue / 100) * perimeter}, ${perimeter}`

      return {
        viewBox,
        radius,
        strokeDasharray,
        perimeter,
        roundValue: `${roundValue}%`,
      }
    })
    return {
      n,
      classes,
      linearProps,
      circleProps,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import './progress';
</style>
