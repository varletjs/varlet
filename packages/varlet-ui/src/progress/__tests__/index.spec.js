import Progress from '..'
import VarProgress from '../Progress'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test progress use', () => {
  const app = createApp({}).use(Progress)
  expect(app.component(Progress.name)).toBeTruthy()
})

describe('test progress component props', () => {
  test('test progress mode', async () => {
    const wrapper = mount(VarProgress, {
      props: { mode: 'linear' },
    })

    expect(wrapper.find('.var-progress__linear').exists()).toBe(true)
    await wrapper.setProps({ mode: 'circle' })
    expect(wrapper.find('.var-progress__circle').exists()).toBe(true)
    wrapper.unmount()
  })

  test('test progress value', () => {
    const wrapper = mount(VarProgress, {
      props: {
        value: 20,
      },
    })

    expect(wrapper.find('.var-progress__linear-certain').attributes('style')).toContain('width: 20%;')
    wrapper.unmount()
  })

  test('test progress line-width', () => {
    const wrapper = mount(VarProgress, {
      props: {
        lineWidth: 5,
      },
    })

    expect(wrapper.find('.var-progress__linear-block').attributes('style')).toContain('height: 5px;')
    wrapper.unmount()
  })

  test('test progress color', () => {
    const wrapper = mount(VarProgress, {
      props: {
        color: 'red',
      },
    })

    expect(wrapper.find('.var-progress__linear-certain').attributes('style')).toContain('background: red;')
    wrapper.unmount()
  })

  test('test progress track-color', () => {
    const wrapper = mount(VarProgress, {
      props: {
        trackColor: 'red',
      },
    })

    expect(wrapper.find('.var-progress__linear-background').attributes('style')).toContain('background: red;')
    wrapper.unmount()
  })

  test('test progress label', async () => {
    const wrapper = mount(VarProgress, {
      props: {
        label: true,
      },
    })

    expect(wrapper.find('.var-progress__linear-label').exists()).toBe(true)
    await wrapper.setProps({ label: false })
    expect(wrapper.find('.var-progress__linear-label').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test progress track', async () => {
    const wrapper = mount(VarProgress, {
      props: {
        track: true,
      },
    })

    expect(wrapper.find('.var-progress__linear-background').exists()).toBe(true)
    await wrapper.setProps({ track: false })
    expect(wrapper.find('.var-progress__linear-background').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test progress ripple', async () => {
    const wrapper = mount(VarProgress, {
      props: {
        ripple: true,
      },
    })

    expect(wrapper.find('.var-progress__linear-ripple').exists()).toBe(true)
    await wrapper.setProps({ ripple: false })
    expect(wrapper.find('.var-progress__linear-ripple').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test progress size', () => {
    const wrapper = mount(VarProgress, {
      props: {
        size: 20,
        mode: 'circle',
      },
    })

    expect(wrapper.find('.var-progress__circle').attributes('style')).toContain('width: 20px; height: 20px;')
    wrapper.unmount()
  })

  test('test progress rotate', () => {
    const wrapper = mount(VarProgress, {
      props: {
        rotate: 10,
        mode: 'circle',
      },
    })

    expect(wrapper.find('.var-progress__circle-svg').attributes('style')).toContain('transform: rotate(-80deg);')
    wrapper.unmount()
  })
})

test('test progress component default slots', () => {
  const wrapper = mount(VarProgress, {
    props: {
      label: true,
    },
    slots: {
      default: () => 'This is default slots',
    },
  })

  expect(wrapper.find('.var-progress__linear-label').element.textContent).toBe('This is default slots')
  wrapper.unmount()
})
