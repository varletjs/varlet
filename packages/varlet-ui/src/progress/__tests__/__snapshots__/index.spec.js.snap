// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test circle progress prop 1`] = `
"<div class=\\"var-progress\\">
  <!--v-if-->
  <div class=\\"var-progress__circle\\" style=\\"width: 50px; height: 50px;\\"><svg class=\\"var-progress__circle-svg\\" style=\\"transform: rotate(-80deg);\\" viewBox=\\"0 0 50 50\\">
      <circle class=\\"var-progress__circle-background\\" cx=\\"25\\" cy=\\"25\\" r=\\"21\\" fill=\\"transparent\\" stroke-width=\\"8\\" style=\\"stroke-dasharray: 131.94689145077132; stroke: #f5cb90;\\"></circle>
      <circle class=\\"var-progress__circle-certain\\" cx=\\"25\\" cy=\\"25\\" r=\\"21\\" fill=\\"transparent\\" stroke-width=\\"8\\" style=\\"stroke-dasharray: 39.5840674352314, 131.94689145077132; stroke: #ff9800;\\"></circle>
    </svg>
    <div class=\\"var-progress__circle-label\\">30%</div>
  </div>
</div>"
`;

exports[`test linear progress prop 1`] = `
"<div class=\\"var-progress\\">
  <div class=\\"var-progress__linear\\">
    <div class=\\"var-progress__linear-block\\" style=\\"height: 8px;\\">
      <div class=\\"var-progress__linear-background\\" style=\\"background: rgb(245, 203, 144);\\"></div>
      <div class=\\"var-progress__linear-certain var-progress__linear-ripple\\" style=\\"background: rgb(255, 152, 0); width: 30%;\\"></div>
    </div>
    <div class=\\"var-progress__linear-label\\">success</div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`test progress example 1`] = `
"<div class=\\"progress-demo\\">
  <div class=\\"space\\"></div>
  <div class=\\"progress-demo__basic\\">
    <div class=\\"app-type\\">基本使用</div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 4px;\\">
          <!--v-if-->
          <div class=\\"var-progress__linear-certain\\" style=\\"width: 20%;\\"></div>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 4px;\\">
          <!--v-if-->
          <div class=\\"var-progress__linear-certain\\" style=\\"width: 0%;\\"></div>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 4px;\\">
          <!--v-if-->
          <div class=\\"var-progress__linear-certain\\" style=\\"width: 100%;\\"></div>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
  </div>
  <div class=\\"progress-demo__label\\">
    <div class=\\"app-type\\">显示标签</div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 4px;\\">
          <div class=\\"var-progress__linear-background\\"></div>
          <div class=\\"var-progress__linear-certain\\" style=\\"width: 30%;\\"></div>
        </div>
        <div class=\\"var-progress__linear-label\\">30%</div>
      </div>
      <!--v-if-->
    </div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 4px;\\">
          <div class=\\"var-progress__linear-background\\"></div>
          <div class=\\"var-progress__linear-certain\\" style=\\"width: 0%;\\"></div>
        </div>
        <div class=\\"var-progress__linear-label\\">0%</div>
      </div>
      <!--v-if-->
    </div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 4px;\\">
          <div class=\\"var-progress__linear-background\\"></div>
          <div class=\\"var-progress__linear-certain\\" style=\\"width: 100%;\\"></div>
        </div>
        <div class=\\"var-progress__linear-label\\">success</div>
      </div>
      <!--v-if-->
    </div>
  </div>
  <div class=\\"progress-demo__style\\">
    <div class=\\"app-type\\">自定义样式</div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 8px;\\">
          <div class=\\"var-progress__linear-background\\"></div>
          <div class=\\"var-progress__linear-certain\\" style=\\"background: rgb(255, 159, 0); width: 30%;\\"></div>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 8px;\\">
          <div class=\\"var-progress__linear-background\\" style=\\"background: rgb(245, 203, 144);\\"></div>
          <div class=\\"var-progress__linear-certain\\" style=\\"background: rgb(255, 159, 0); width: 60%;\\"></div>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div class=\\"var-progress\\">
      <div class=\\"var-progress__linear\\">
        <div class=\\"var-progress__linear-block\\" style=\\"height: 8px;\\">
          <div class=\\"var-progress__linear-background\\" style=\\"background: rgb(245, 203, 144);\\"></div>
          <div class=\\"var-progress__linear-certain var-progress__linear-ripple\\" style=\\"background: rgb(255, 159, 0); width: 80%;\\"></div>
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
  </div>
  <div class=\\"app-type\\">环形进度条</div>
  <div class=\\"progress-demo__circle\\">
    <div class=\\"var-progress\\">
      <!--v-if-->
      <div class=\\"var-progress__circle\\" style=\\"width: 56px; height: 56px;\\"><svg class=\\"var-progress__circle-svg\\" style=\\"transform: rotate(-90deg);\\" viewBox=\\"0 0 56 56\\">
          <circle class=\\"var-progress__circle-background\\" cx=\\"28\\" cy=\\"28\\" r=\\"25.5\\" fill=\\"transparent\\" stroke-width=\\"5\\" style=\\"stroke-dasharray: 160.22122533307945;\\"></circle>
          <circle class=\\"var-progress__circle-certain\\" cx=\\"28\\" cy=\\"28\\" r=\\"25.5\\" fill=\\"transparent\\" stroke-width=\\"5\\" style=\\"stroke-dasharray: 48.06636759992383, 160.22122533307945;\\"></circle>
        </svg>
        <!--v-if-->
      </div>
    </div>
    <div class=\\"space\\"></div>
    <div class=\\"var-progress\\">
      <!--v-if-->
      <div class=\\"var-progress__circle\\" style=\\"width: 56px; height: 56px;\\"><svg class=\\"var-progress__circle-svg\\" style=\\"transform: rotate(-90deg);\\" viewBox=\\"0 0 56 56\\">
          <circle class=\\"var-progress__circle-background\\" cx=\\"28\\" cy=\\"28\\" r=\\"25.5\\" fill=\\"transparent\\" stroke-width=\\"5\\" style=\\"stroke-dasharray: 160.22122533307945;\\"></circle>
          <circle class=\\"var-progress__circle-certain\\" cx=\\"28\\" cy=\\"28\\" r=\\"25.5\\" fill=\\"transparent\\" stroke-width=\\"5\\" style=\\"stroke-dasharray: 0, 160.22122533307945;\\"></circle>
        </svg>
        <div class=\\"var-progress__circle-label\\">0%</div>
      </div>
    </div>
    <div class=\\"space\\"></div>
    <div class=\\"var-progress\\">
      <!--v-if-->
      <div class=\\"var-progress__circle\\" style=\\"width: 56px; height: 56px;\\"><svg class=\\"var-progress__circle-svg\\" style=\\"transform: rotate(-90deg);\\" viewBox=\\"0 0 56 56\\">
          <circle class=\\"var-progress__circle-background\\" cx=\\"28\\" cy=\\"28\\" r=\\"25.5\\" fill=\\"transparent\\" stroke-width=\\"5\\" style=\\"stroke-dasharray: 160.22122533307945;\\"></circle>
          <circle class=\\"var-progress__circle-certain\\" cx=\\"28\\" cy=\\"28\\" r=\\"25.5\\" fill=\\"transparent\\" stroke-width=\\"5\\" style=\\"stroke-dasharray: 160.22122533307945, 160.22122533307945;\\"></circle>
        </svg>
        <div class=\\"var-progress__circle-label\\">100%</div>
      </div>
    </div>
  </div>
  <div class=\\"app-type\\">隐藏轨道</div>
  <div class=\\"progress-demo__circle\\">
    <div class=\\"var-progress\\">
      <!--v-if-->
      <div class=\\"var-progress__circle\\" style=\\"width: 56px; height: 56px;\\"><svg class=\\"var-progress__circle-svg\\" style=\\"transform: rotate(-90deg);\\" viewBox=\\"0 0 56 56\\">
          <!--v-if-->
          <circle class=\\"var-progress__circle-certain\\" cx=\\"28\\" cy=\\"28\\" r=\\"26\\" fill=\\"transparent\\" stroke-width=\\"4\\" style=\\"stroke-dasharray: 81.68140899333463, 163.36281798666926;\\"></circle>
        </svg>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;
